events/sinks/kafka/driver.go

```diff
@@ -0,0 +1,175 @@
+	brokerDialRetryWait    = 0
+	brokerLeaderRetryLimit = 1
+	brokerLeaderRetryWait  = 0
+	dataTopic              = "heapster-events"
+)
+
+type KafkaSinkPoint struct {
+	EventValue     interface{}
+	EventTimestamp time.Time
+	EventTags      map[string]string
+}
+
+type kafkaSink struct {
+	producer  kafka.Producer
+	dataTopic string
+	sync.RWMutex
+}
+
+func getEventValue(event *kube_api.Event) (string, error) {
+	// TODO: check whether indenting is required.
+	bytes, err := json.MarshalIndent(event, "", " ")
+	if err != nil {
+		return "", err
+	}
+	return string(bytes), nil
+}
+
+func eventToPoint(event *kube_api.Event) (*KafkaSinkPoint, error) {
+	value, err := getEventValue(event)
+	if err != nil {
+		return nil, err
+	}
+	point := KafkaSinkPoint{
+		EventTimestamp: event.LastTimestamp.Time.UTC(),
+		EventValue:     value,
+		EventTags: map[string]string{
+			"eventID": string(event.UID),
+		},
+	}
+	if event.InvolvedObject.Kind == "Pod" {
+		point.EventTags[core.LabelPodId.Key] = string(event.InvolvedObject.UID)
+		point.EventTags[core.LabelPodName.Key] = event.InvolvedObject.Name
+	}
+	point.EventTags[core.LabelHostname.Key] = event.Source.Host
+	return &point, nil
+}
+
+func (sink *kafkaSink) ExportEvents(eventBatch *event_core.EventBatch) {
+	sink.Lock()
+	defer sink.Unlock()
+
+	for _, event := range eventBatch.Events {
+		point, err := eventToPoint(event)
+		if err != nil {
+			glog.Warningf("Failed to convert event to point: %v", err)
+		}
+		sink.produceKafkaMessage(*point, sink.dataTopic)
+	}
+}
+
+func (sink *kafkaSink) produceKafkaMessage(dataPoint KafkaSinkPoint, topic string) error {
+	start := time.Now()
+	jsonItems, err := json.Marshal(dataPoint)
+	if err != nil {
+		return fmt.Errorf("failed to transform the items to json : %s", err)
+	}
+	message := &proto.Message{Value: []byte(string(jsonItems))}
+	_, err = sink.producer.Produce(topic, partition, message)
+	if err != nil {
+		return fmt.Errorf("failed to produce message to %s:%d: %s", topic, partition, err)
+	}
+	end := time.Now()
+	glog.V(4).Info("Exported %d data to kafka in %s", len([]byte(string(jsonItems))), end.Sub(start))
+	return nil
+}
+
+func (sink *kafkaSink) Name() string {
+	return "Apache Kafka Sink"
+}
+
+func (sink *kafkaSink) Stop() {
+	// nothing needs to be done.
+}
+
+// setupProducer returns a producer of kafka server
+func setupProducer(sinkBrokerHosts []string, brokerConf kafka.BrokerConf) (kafka.Producer, error) {
+	glog.V(3).Infof("attempting to setup kafka sink")
+	broker, err := kafka.Dial(sinkBrokerHosts, brokerConf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to kafka cluster: %s", err)
+	}
+	defer broker.Close()
+
+	//create kafka producer
+	conf := kafka.NewProducerConf()
+	conf.RequiredAcks = proto.RequiredAcksLocal
+	sinkProducer := broker.Producer(conf)
+	glog.V(3).Infof("kafka sink setup successfully")
+	return sinkProducer, nil
+}
+
+func NewKafkaSink(uri *url.URL) (event_core.EventSink, error) {
```

Comment:
moving this logic of initing action to common directory; https://github.com/kubernetes/heapster/tree/master/common